fnlb/main.go

```diff
@@ -64,27 +72,57 @@ func main() {
 		return r.URL.Path, nil
 	}
 
-	h := lb.NewProxy(k, g, r, conf)
-	h = g.Wrap(h) // add/del/list endpoints
-	h = r.Wrap(h) // stats / dash endpoint
+	servers := make([]*http.Server, 0, 1)
+	handler := lb.NewProxy(k, g, r, conf)
 
-	err = serve(conf.Listen, h)
-	if err != nil {
-		logrus.WithError(err).Fatal("server error")
+	// a separate mgmt listener is requested? then let's create a LB traffic only server
+	if conf.Listen != conf.MgmtListen {
+		servers = append(servers, &http.Server{Addr: conf.Listen, Handler: handler})
+		handler = lb.NullHandler()
 	}
+
+	// add mgmt endpoints to the handler
+	handler = g.Wrap(handler) // add/del/list endpoints
+	handler = r.Wrap(handler) // stats / dash endpoint
+
+	servers = append(servers, &http.Server{Addr: conf.MgmtListen, Handler: handler})
+	serve(servers, &conf)
 }
 
-func serve(addr string, handler http.Handler) error {
-	server := &http.Server{Addr: addr, Handler: handler}
+func serve(servers []*http.Server, conf *lb.Config) {
 
 	ch := make(chan os.Signal, 1)
 	signal.Notify(ch, syscall.SIGQUIT, syscall.SIGINT)
-	go func() {
-		for sig := range ch {
-			logrus.WithFields(logrus.Fields{"signal": sig}).Info("received signal")
-			server.Shutdown(context.Background()) // safe shutdown
-			return
+
+	for i := 0; i < len(servers); i++ {
+		go func(idx int) {
+			err := servers[idx].ListenAndServe()
+			if err != nil && err != http.ErrServerClosed {
+				logrus.WithFields(logrus.Fields{"server_id": idx}).WithError(err).Fatal("server error")
+			} else {
+				logrus.WithFields(logrus.Fields{"server_id": idx}).Info("server stopped")
+			}
+		}(i)
+	}
+
+	sig := <-ch
+	logrus.WithFields(logrus.Fields{"signal": sig}).Info("received signal")
+
+	for i := 0; i < len(servers); i++ {
+
+		ctx := context.Background()
+
+		if conf.ShutdownTimeout > 0 {
+			tmpCtx, cancel := context.WithTimeout(context.Background(), time.Duration(conf.ShutdownTimeout)*time.Second)
+			ctx = tmpCtx
+			defer cancel()
```

Comment:
super nit: `ctx, _ = context.WithTimeout(ctx, ...)` and removing var assign / cancel seems fine here since we're shutting down, not too worried about context leaking.

Reply:
"go vet" complains about leaked context... I'll retain 'defer cancel()' if that's ok.